### PR TITLE
ceph docs: identify required ceph configs

### PR DESCRIPTION
**Pull request by BlaineEXE**
Prepare user for 'required' settings earlier in the upgrade doc.

Signed-off-by: Blaine Gardner <blaine.gardner@suse.com>

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [x] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[skip ci]

### CODE REVIEW

File: pkg/operator/ceph/config/config.go [hunk 1]
```diff
@@ -133,6 +133,57 @@ func SetDefaultAndUserConfigs(
...
+	configOverrides rookceph.ConfigOverridesSpec,
+) error {
+	// ceph.conf is never used. All configurations are made in the centralized mon config database,
+	// or they are specified on the commandline when daemons are called.
+	monStore := GetMonStore(context, namespace)
+
+	// one-time defaults
+	defaults := map[string]string{
+		"osd pool default pg num":  "128",
+		"osd pool default pgp num": "128",
```
[the remark this answers is not in the document]
Comment (mbukatov) [reply pkg/operator/ceph/config/config.go]: There are some hints about selecting a pg number in ceph docs: [A preselection of pg_num](https://docs.ceph.com/docs/master/rados/operations/placement-groups/#a-preselection-of-pg-num).

And it seems that there may be a simple answer to the pg number, but for small clusters only:

> If you have more than 50 OSDs, you need to understand the tradeoffs and how to calculate the pg_num value by yourself

But even for small clusters, pg num set to 100 is not a good one:

> pg number should always be rounded up to the nearest power of two

[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/config/config.go]: IIRC Octopus should enable the PG autocaler by default on new clusters. I'm **really** against changing Ceph's default values of `osd_pool_default_pg_num` and `osd_pool_default_pgp_num`. I believe, by now, it has been well understood that PGs must be configured appropriately according to your cluster design (amount of OSDs). Changing the value to 128 is not helping.

Our CI should set both values in accordance to the CI cluster test so if 128 is enough then configure the Test suite to deploy a cluster with 128. 

For everything else, user should be aware that this has to be configured. So this is mostly doc for me.

To sum up, I'm against re-introducing old values, as well as adding new once. I think what we have now is correct.

[the remark this answers is not in the document]
Comment (travisn) [reply pkg/operator/ceph/config/config.go]: @leseb @mbukatov @liewegas Can we enable the PG autoscaler for new clusters now? Or do we need to wait for octopus? 

If we do enable it by default now, I would suggest upgraded clusters by default not have it enabled. They could enable a setting in the CRD later when they are ready to enable it.

But if we shouldn't enable the autoscaler until octopus, it seems safer to keep the defaults Rook already has been using rather than introduce new defaults.

[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/config/config.go]: I'd wait for whatever Ceph does for new clusters but maybe @liewegas has a different option.

I don't think we need a CRD option for that, they can simply go on the toolbox, what'd be nice though is to have the ability to configure MGR modules via the CRD but that's a different discussion.

I still see re-introducing "wrong" values as invalid, even though it hasn't broken anything AFAIK, but since this is the master branch I wouldn't do anything as mentioned above.

File: Documentation/ceph-configuration.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
+---
+title: Configuration
+weight: 3700
```
Comment (travisn) [inline Documentation/ceph-configuration.md]: This topic will come between the CSI and upgrades topics?

File: Documentation/ceph-configuration.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+detailed [here](ceph-toolbox.md). From here, the user can change Ceph configurations, enable manager
+modules, create users and pools, and much more.
+
+### Ceph Dashboard
+The Ceph Dashboard, examined in more detail [here](ceph-dashboard.md) is another way of setting
+some of Ceph's configuration directly. Configuration by the Ceph dashboard is recommended with the
+same priority as configuration via the Ceph CLI (above)
+
+### Config override in CephCluster resource
+Some users may want to specify some settings they want set in the `CephCluster` manifest. This is
```
Comment (travisn) [inline Documentation/ceph-configuration.md]: @leseb and I aren't seeing the need to expose it in the CR. Having them use the override ConfigMap seems sufficient for advanced users. Do we know a scenario where the other options aren't sufficient? I'd rather keep it simple than expose the complexity in the CR.

File: Documentation/ceph-configuration.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+---
+title: Configuration
+weight: 3700
+indent: true
+---
+
+# Configuration
+
+For most any Ceph cluster, the user will want to -- and may need to -- change some Ceph
```
Comment (travisn) [inline Documentation/ceph-configuration.md]: `--` can be filled in?

File: cluster/examples/kubernetes/ceph/cluster.yaml [hunk 5]
```diff
@@ -145,16 +145,25 @@ spec:
...
     - who: global
-      option: debug ms
-      value: "1/5"
-    - who: osd.0
-      option: debug_osd
-      value: "10"
-    - who: mon.a
-      option: debug-mon
-      value: "10"
+      option: osd_pool_default_pg_num
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/cluster.yaml]: I thought we agreed on keeping Ceph's default, thus not exposing `osd_pool_default_pg_num` nor `osd_pool_default_pgp_num`. Did I miss something?

File: Documentation/ceph-configuration.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+[Ceph configuration documentation](https://docs.ceph.com/docs/master/rados/configuration/) before
+making any changes. Changing the settings could result in unhealthy daemons or even data loss if
+used incorrectly.
+
+## Required configurations
+Rook and Ceph both strive to make configuration as easy as possible, but there are some
+configuration options which users are well advised to consider for any production cluster.
+
+### Default PG and PGP counts
+`osd_pool_default_pg_num` and `osd_pool_default_pgp_num`. The number of PGs and PGPs can be
```
Comment (travisn) [inline Documentation/ceph-configuration.md]: Users are going to see issues with PGs surfaced as warnings so I wonder if they will look to the troubleshooting guide first in order to resolve the warnings. What if we put PG configuration in ceph-common-issues.md, and reserve this for the more general topic of how to apply other ceph config options? Or else we keep this topic here and promote it to a larger title so they see that they need to learn how to configure PGs.

File: cluster/examples/kubernetes/ceph/cluster.yaml (hunk 5 above)
Comment (BlaineEXE) [reply cluster/examples/kubernetes/ceph/cluster.yaml]: This is part of the doc updates to indicate to users that they should pay attention to these  configs. We could comment them out if that seems better.

File: Documentation/ceph-configuration.md (hunk 3 above)
Comment (leseb) [reply Documentation/ceph-configuration.md]: Yes, as discussed, I think overall we introduced a lot more complex than what we used to have the override configmap. I'm 100% in favor of getting rid of any `ceph.conf` don't get me wrong, however, for specific overrides, a `ceph.conf` is perfectly valid. Exposing the mon config store through the CRD is difficult to maintain and to honor. At its best, the CR can honor additions and updates, but won't be able to detect deletion and revert to the previous value or reset it to its default. Using the CR as a source of truth really adds a lot of complexity and in the end, I only see users being confused, not understanding what's going on.

File: Documentation/ceph-configuration.md (hunk 2 above)
Comment (BlaineEXE) [reply Documentation/ceph-configuration.md]: That was the intent from looking at the current docs on the website. I'm open to other suggestions.

File: Documentation/ceph-configuration.md (hunk 3 above)
Comment (BlaineEXE) [reply Documentation/ceph-configuration.md]: This is not the PR to comment about this issue. Here is where this input is needed. https://github.com/rook/rook/pull/3767

File: Documentation/ceph-configuration.md (hunk 4 above)
Comment (BlaineEXE) [reply Documentation/ceph-configuration.md]: I don't understand. Does this render poorly?

File: Documentation/ceph-configuration.md (hunk 6 above)
Comment (BlaineEXE) [reply Documentation/ceph-configuration.md]: Waiting until the users might notice an error doesn't seem like the right time to confront this topic. Configuration is a necessary part of any software, and it should be a prominent topic in Rook's docs. I'm not entirely sure I understand what you are suggesting by promoting it to a larger title.

File: cluster/examples/kubernetes/ceph/cluster.yaml (hunk 5 above)
Comment (BlaineEXE) [reply cluster/examples/kubernetes/ceph/cluster.yaml]: @leseb @travisn I would like some thoughts about this.

File: Documentation/ceph-upgrade.md [hunk 7]
```diff
@@ -60,6 +60,12 @@ any time without compatibility support and without prior notice.
 was the last Rook release which will support Ceph's Luminous (v12.x.x) version. These are the only
 supported major versions of Ceph.
 
+Rook documentation for 1.1 has identified some Ceph configuration options that the user is highly
+advised to consider. See the docs [here](ceph-configuration.md#default-pg-and-pgp-counts). It is
+highly advised to consider these now before upgrading Rook as well. If the user determines that
```
Comment (travisn) [inline Documentation/ceph-upgrade.md]: - nit: "highly advised" is mentioned twice. Do we need to be that alarming?
- I wonder if this is better to recommend as a post-upgrade consideration. At that point they will be able to use the new cluster CR option for enabling the pg autoscaler.

Comment (BlaineEXE) [reply Documentation/ceph-upgrade.md]: It is a post-upgrade consideration, but it's mentioned here to prepare the user. I should probably update this to mention the mgr module enabling options.
